### PR TITLE
Allow crate users to define custom models - alternative using generics

### PR DESCRIPTION
**Pull request by uint**
This
* makes `Simulation` generic over any type that implements `AsModel`,
* provides a derive macro to conveniently implement custom enums that encompass possible model types (sort of like how we used `enum_dispatch`),
* removes `enum_dispatch` in favor of the above,
* provides an example of adding a custom model in `sim/tests/custom_models.rs`.

### CODE REVIEW

**Comment by uint** [comment]
The wins:
* In theory, this should incur no extra overhead over what we had before this PR - we're only using zero-cost abstractions.
* More idiomatic - there's (IMO) less unexpected stuff for new contributors to mull over.
* Encoding information in Rust's type system rather than runtime generally leaves less room for error.

Next steps:
* Possibly tests for the proc macro? It's very simple though.
* A more thorough test in `custom_models.rs` - right now it doesn't test correctness, it just makes sure things compile with the derive macro used.
* Since we're separating crates now, consider putting `WebSimulation` in a separate crate? `sim_wasm` sounds about right.

**Comment by ndebuhr** [comment]
Awesome!  Thanks @uint.  How should a `WebSimulation` test (with custom models) look?  In the code below, I don't know how to use the `ModelType` in this situation.  Can you help me with this?  Or is this use case not covered by the solution?  To experiment, copy-paste that code into a new integration test file.  Also, is the `smart-default` essential to the solution?  The last update was March 2020.  I would rate that repository health as below average, but it's probably ok if we have the option to refactor this out later.  https://github.com/idanarye/rust-smart-default.

```rust
use std::f64::INFINITY;

use serde::{Deserialize, Serialize};
use sim::models::*;
use sim::simulator::*;
use sim::SimulationError;
use wasm_bindgen::prelude::*;
use wasm_bindgen_test::*;

/// The passive model does nothing
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Passive {
    ports_in: PortsIn,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct PortsIn {
    job: String,
}

impl Passive {
    pub fn new(job_port: String) -> Self {
        Self {
            ports_in: PortsIn { job: job_port },
        }
    }
}

impl AsModel for Passive {
    fn status(&self) -> String {
        "Passive".into()
    }

    fn events_ext(
        &mut self,
        _incoming_message: ModelMessage,
        _services: &mut Services,
    ) -> Result<Vec<ModelMessage>, SimulationError> {
        Ok(Vec::new())
    }

    fn events_int(
        &mut self,
        _services: &mut Services,
    ) -> Result<Vec<ModelMessage>, SimulationError> {
        Ok(Vec::new())
    }

    fn time_advance(&mut self, _time_delta: f64) {
        // No future events list to advance
    }

    fn until_next_event(&self) -> f64 {
        // No future events list, as a source of finite until_next_event
        // values
        INFINITY
    }
}

#[derive(Deserialize, Serialize, Clone, Debug, AsModel)]
pub enum ModelType {
    // Built-in models used
    Generator(Generator),
    // Custom models used
    Passive(Passive),
}

#[wasm_bindgen_test]
fn step_n_with_custom_passive_model_wasm() {
    let models = r#"
- type: "Generator"
  id: "generator-01"
  portsIn: {}
  portsOut:
    job: "job"
  messageInterdepartureTime:
    exp:
      lambda: 0.5
- type: "Passive"
  id: "passive-01"
  portsIn:
    job: "job"
"#;
    let connectors = r#"
- id: "connector-01"
  sourceID: "generator-01"
  targetID: "passive-01"
  sourcePort: "job"
  targetPort: "job"
"#;

    let mut simulation = WebSimulation::post_yaml(&models, &connectors);
    // 1 initialization event, and 2 events per generation
    let messages: Vec<Message> = serde_json::from_str(&simulation.step_n_json(9)).unwrap();
    let generations_count = messages.len();
    let expected = 4; // 4 interarrivals from 9 steps
    assert_eq!(generations_count, expected);
}
```

**Comment by uint** [comment]
@ndebuhr Sorry for the wordiness!

> How should a `WebSimulation` test (with custom models) look? In the code below, I don't know how to use the `ModelType` in this situation. Can you help me with this? Or is this use case not covered by the solution?

I didn't cover making `WebSimulation` work with custom models, only `Simulation`. It isn't possible to make a thing we export for use as a WASM lib generic, it seems.

I'm thinking of it this way though: what use case would this solve? If there's a user wanting to use the crate as a dependency in their Rust project, why would they want to use the `WebSimulation` struct specifically?

Personally, I'd move WebSimulation into its own crate - something like `sim_wasm`, just to separate the concerns here. If someone wants to build their own version of the WASM lib, they can fork `sim_wasm` and depend on the core `sim` library, which will then be light and flexible.

> Also, is the `smart-default` essential to the solution? The last update was March 2020. I would rate that repository health as below average, but it's probably ok if we have the option to refactor this out later. https://github.com/idanarye/rust-smart-default.

I stumbled into a problem when working on the PR: https://github.com/rust-lang/rust/issues/26925

In short, `#derive[Default]` from the stdlib incorrectly requires that all type parameters of the struct also implement `Default`. In our case it expects `M: AsModel` will implement `Default` even though it's not really needed. `Vec` will default to an empty vector, so whether the underlying type has a default is unimportant.

```rust
pub struct Simulation<M: AsModel> {
    models: Vec<Model<M>>,
    ..
}
```

`smart_default` does a better job here. That crate is small and probably just doesn't need much updating, but I could review its code to make sure it's okay. We could contribute upstream if necessary. I think it beats writing a custom implementation of `Default` for `Simulation`.

**Comment by ndebuhr** [comment]
@uint 
> I'm thinking of it this way though: what use case would this solve? If there's a user wanting to use the crate as a dependency in their Rust project, why would they want to use the `WebSimulation` struct specifically?

The main use case is _someone has a simulation problem that is best solved with a web application_.  The user would:
1. Create a Rust project with a `sim` dependency and custom models
2. Compile to wasm and package/publish with npm
3. Use that npm package in a web application

`Simulation` is not (and currently can't be) wasm-bindgen exported like `WebSimulation`.  I know it's a fundamentally different approach, with its own benefits and drawbacks, but https://github.com/ndebuhr/sim/pull/8 does support custom models for `WebSimulation`.  Is there any way to combine and get the best of both solutions?  It may look like a rare use case, but personally, I have already seen `WebSimulation` custom model needs.  I think that allowing `WebSimulation` custom models is worth a performance and complexity cost.

What are your thoughts?

**Comment by uint** [comment]
> @uint
> 
> The main use case is _someone has a simulation problem that is best solved with a web application_. The user would:
> 
>     1. Create a Rust project with a `sim` dependency and custom models
> 
>     2. Compile to wasm and package/publish with npm
> 
>     3. Use that npm package in a web application
> 
This is possible. It's just that the user has to define their own bindgen API, but I suspect they would have to even with the other PR. I worry that if they create their own crate called `simul` that depends on `sim`, this `simul` crate won't be able to re-export `WebSimulation` (or its methods) to be used by JS. They would still end up creating their own version of `WebSimulation` from scratch and doing bindgen for it, wouldn't they?

Am I understanding `wasm_bindgen_test` correctly? I think it's just unit tests that run in the browser's WASM environment. It doesn't really test the API you export with wasm_bindgen, right?

I've been trying to confirm my suspicions with an example JS thingy, but I'm struggling a little with this in my environment currently.

> `Simulation` is not (and currently can't be) wasm-bindgen exported like `WebSimulation`. I know it's a fundamentally different approach, with its own benefits and drawbacks, but #8 does support custom models for `WebSimulation`.

For the record, I don't think `Simulation` should be wasm-bindgen. I think the exported API should be thin and focused.

> Is there any way to combine and get the best of both solutions?

There is always the possibility of doing generics like in this PR, but also making sure `Box<dyn AsModel>` can be used in a `Simulation<Box<dyn AsModel>>`. That sounds sane to me. The user can choose their own tradeoffs with `Simulation` then.

It would also be possible to create two versions of `WebSimulaton` - one with the trait object and one with the built-in enum to let the user choose the tradeoffs they want, though I'm not sure I like this idea all that much.

Would it be useful to discuss this on a video call?

**Comment by ndebuhr** [comment]
@uint I agree on most things.

> This is possible. It's just that the user has to define their own bindgen API, but I suspect they would have to even with the other PR. I worry that if they create their own crate called `simul` that depends on `sim`, this `simul` crate won't be able to re-export `WebSimulation` (or its methods) to be used by JS. They would still end up creating their own version of `WebSimulation` from scratch and doing bindgen for it, wouldn't they?

We do have coverage of this in https://github.com/ndebuhr/sim/pull/8, without requiring the crate user to redefine `WebSimulation` or bindgen APIs.  Something like the code below (as a `simul` separate Rust project).  I don't completely understand the mechanics of what's happening, but am experimenting and can verify that the `simul` code below, compiled to wasm, can be used in npm projects (including the custom `Passive` model).  In JavaScript, the `DynamicsSimulation::post_json` allows for passing `Passive` model specifications (and the other "out-of-the-box" `sim` models).
```rust
use sim::simulator::WebSimulation;
use sim_derive::{register};
use wasm_bindgen::prelude::*;

use crate::utils;

mod passive;
pub use passive::Passive;

#[wasm_bindgen]
pub struct DynamicsSimulation {}

#[wasm_bindgen]
impl DynamicsSimulation {
    pub fn post_json(nodes: &str, connectors: &str) -> WebSimulation {
        utils::set_panic_hook();
        register![Passive];
        WebSimulation::post_json(nodes, connectors)
    }

    pub fn post_yaml(nodes: &str, connectors: &str) -> WebSimulation {
        utils::set_panic_hook();
        register![Passive];
        WebSimulation::post_yaml(nodes, connectors)
    }
}
```
and the Passive custom model definition:
```rust
use std::f64::INFINITY;

use serde::{Deserialize, Serialize};
use sim::models::model_trait::{AsModel, SerializableModel};
use sim::models::ModelMessage;
use sim::simulator::Services;
use sim::utils::error::SimulationError;
use sim_derive::SerializableModel;

/// The passive model does nothing
#[derive(Debug, Clone, Serialize, Deserialize, SerializableModel)]
#[serde(rename_all = "camelCase")]
pub struct Passive {
    ports_in: PortsIn,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct PortsIn {
    job: String,
}

impl Passive {
    pub fn new(job_port: String) -> Self {
        Self {
            ports_in: PortsIn { job: job_port },
        }
    }
}

impl AsModel for Passive {
    fn status(&self) -> String {
        "Passive".into()
    }

    fn events_ext(
        &mut self,
        _incoming_message: ModelMessage,
        _services: &mut Services,
    ) -> Result<Vec<ModelMessage>, SimulationError> {
        Ok(Vec::new())
    }

    fn events_int(
        &mut self,
        _services: &mut Services,
    ) -> Result<Vec<ModelMessage>, SimulationError> {
        Ok(Vec::new())
    }

    fn time_advance(&mut self, _time_delta: f64) {
        // No future events list to advance
    }

    fn until_next_event(&self) -> f64 {
        // No future events list, as a source of finite until_next_event
        // values
        INFINITY
    }
}
```
I would prefer to keep the rationale documented and public, if possible.  GitHub comment conversations are preferred, but if that takes too much time, a video call could be good.

https://github.com/ndebuhr/sim/pull/8 seems to have maximum use case coverage, including the:
1. Create a Rust project with a `sim` dependency and custom models
2. Compile to wasm and package/publish with npm
3. Use that npm package in a web application

_If we need to make tradeoffs_, I think the coverage of this important use case and the good ergonomics outweigh performance, `sim` complexity, and `sim` idiomatic-ness drawbacks.  

@uint What do you think?

**Comment by ndebuhr** [comment]
@uint I merged https://github.com/ndebuhr/sim/pull/8 and will work on good automated tests.  After that, we will be in a good position to evaluate and benchmark alternative solutions, and to easily understand use case coverage.
